src/raftstore/store/peer.rs

```diff
@@ -606,87 +629,86 @@ impl Peer {
 
         self.update_leader_lease(&ready);
 
-        self.add_ready_metric(&ready, &mut metrics.ready);
+        let ready_metrics = ctx.metrics.ready.clone();
+        self.add_ready_metric(&ready, &mut ctx.metrics.ready);
 
         // The leader can write to disk and replicate to the followers concurrently
         // For more details, check raft thesis 10.2.1.
         if self.is_leader() {
-            self.send(trans, ready.messages.drain(..), &mut metrics.message)
-                .unwrap_or_else(|e| {
-                    // We don't care that the message is sent failed, so here just log this error.
-                    warn!("{} leader send messages err {:?}", self.tag, e);
-                })
+            let msgs = ready.messages.drain(..);
+            self.send(ctx.trans, msgs, &mut ctx.metrics.message).unwrap_or_else(|e| {
+                // We don't care that the message is sent failed, so here just log this error.
+                warn!("{} leader send messages err {:?}", self.tag, e);
+            });
         }
 
         let append_timer = PEER_APPEND_LOG_HISTOGRAM.start_timer();
-        let apply_result = match self.mut_store().handle_raft_ready(&ready) {
+        let invoke_ctx = match self.mut_store().handle_raft_ready(ctx, &ready) {
             Ok(r) => r,
             Err(e) => {
-                // Leader has to panic when failed to persist raft entries,
-                // because all the MsgAppend messages have been sent out already.
-                // Suppose there are 2 followers. If any follower sends back a
-                // MsgAppendResponse, the leader will commit the non-persistent
-                // entries.
-                if self.is_leader() {
-                    panic!("{} failed to handle raft ready: {:?}", self.tag, e);
-                }
-                return Err(e);
+                // We may have written something to writebatch and it can't be reverted, so has
+                // to panic here.
+                panic!("{} failed to handle raft ready: {:?}", self.tag, e);
             }
         };
         append_timer.observe_duration();
 
-        if !self.is_leader() {
-            self.send(trans, ready.messages.drain(..), &mut metrics.message)
-                .unwrap_or_else(|e| {
-                    warn!("{} follower send messages err {:?}", self.tag, e);
-                })
-        }
-
         slow_log!(t,
```

Comment:
slow log seems no meaning here now. We may add slow log when writing the append WriteBatch.